### ISSUE TITLE
Direct3D12 UWP Game DR (C++/WinRT) template generated code crashes

### ISSUE THREAD

**Issue by anup-das** [walbourn/directx-vs-templates#20]
Current system configuration:
Windows 10 17763.134
Visual Studio 2017 15.9.2
Direct3DUWPGame  VSIX version  6.3
Graphics Intel HD 4600 and NVIDIA GeForce GT 755M (optimus technology) with latest drivers from Intel and NVIDIA.

Note: I have been observing this error since  very early days of Creator's update. I thought by this time this issue will be resolved but it hasn't.

I compile and run the default generated code in 32-bit debug build and resize the window continuously until it crashes and observe the following messages in output window

[log.txt](https://github.com/walbourn/directx-vs-templates/files/2606169/log.txt)

DXGI ERROR: IDXGISwapChain::GetContainingOutput: The swapchain's adapter does not control the output on which the swapchain's window resides. [ MISCELLANEOUS ERROR #80: ]
... repeated message above too many times
DXGI ERROR: IDXGISwapChain::ResizeBuffers: Swapchain cannot be resized unless all outstanding buffer references have been released. [ MISCELLANEOUS ERROR #19: ]
Exception thrown at 0x76BAAB32 in Direct3D UWP Game1.exe: Microsoft C++ exception: DX::com_exception at memory location 0x0314CE90.
Exception thrown at 0x76BAAB32 in Direct3D UWP Game1.exe: Microsoft C++ exception: [rethrow] at memory location 0x00000000.
Exception thrown at 0x76BAAB32 (KernelBase.dll) in Direct3D UWP Game1.exe: WinRT originate error - 0x80004005 : 'Failure with HRESULT of 887A0001'.

I see the DXGI Error message with Direct3D 11 but for some reason it does not seem to crash. Direct3D12 code crashes easily.

**Comment by walbourn** [comment]
Which Windows 10 SDK are you using? I'm not able to repro this on any of my systems.

The only place I use ``GetContainingOutput`` is in the DeviceReources function ``UpdateColorSpace`` which I handle failure gracefully, so I think that debug output might be noise.

**Comment by walbourn** [comment]
The DeviceResources implementation is using the new [IDXGIFactory6::EnumAdapterByGpuPreference](https://docs.microsoft.com/en-us/windows/desktop/api/dxgi1_6/nf-dxgi1_6-idxgifactory6-enumadapterbygpupreference) method to prefer the NVIDIA over the Intel on your system. You can try commenting out that code block in your DR to see if it avoids the problem.

**Comment by anup-das** [comment]
OK I modified the function void DeviceResources::CreateDeviceResources() to comment the call to GetAdapter and pass nullptr directly to D3D12CreateDevice() as follows..

    ComPtr<IDXGIAdapter1> adapter;
    //GetAdapter(adapter.GetAddressOf());

    // Create the DX12 API device object.
    ThrowIfFailed(D3D12CreateDevice(
        nullptr, //adapter.Get(),
        m_d3dMinFeatureLevel,
        IID_PPV_ARGS(m_d3dDevice.ReleaseAndGetAddressOf())
        ));
With this change the app selects GPU 0  i.e. Intel, the resizing horribly and exceptionally slow with too much lag but it never crashes.  Now I tried to override the adapter by going to settings and selecting high performance (See attached screenshot) and I still have a crash. See log.

![default adapter crash](https://user-images.githubusercontent.com/42484400/49038733-caf93500-f1e3-11e8-94c8-ac03ca0d47fc.png)

[Default Adapter Crash.txt](https://github.com/walbourn/directx-vs-templates/files/2617128/Default.Adapter.Crash.txt)

**Comment by walbourn** [comment]
What Windows 10 SDK are you building against?

**Comment by anup-das** [comment]
![target version](https://user-images.githubusercontent.com/42484400/49038943-625e8800-f1e4-11e8-911b-7635c396a11e.png)

**Comment by walbourn** [comment]
So the core problem here is that the [](https://docs.microsoft.com/en-us/uwp/api/windows.ui.core.corewindow.resizestarted) and []() events have regressed. I'm no longer getting the events, so I'm not able to avoid the dozens of resizes per second. I've asked for investigation on this from the Windows team.

The ``GetContainingOutput`` message is just noise that should be suppressed, so I've added that.

**Comment by anup-das** [comment]
As initially commented I am seeing this issue since Creator's update. Also why does this happen only with DX12 ?  
Anyway let me know if you want me to do any experimentation stuff at my end.  I am currently sticking to DX11 for my UWP app development for simplicity but in future I might want to use DX12.

**Comment by walbourn** [comment]
The _ResizeStarted_ event behavior is missing for both DX11 & DX 12, which is the direct solution for the 'slow resize' problem. This was apparently regressed in the Windows 10 (October 2018; build 17763) release, so they are looking at that.

Now as to the problem you are seeing with _ResizeBuffers_ failing, I'm beginning to think that is an NVIDIA bug. I can't reproduce it on any system other than an NVIDIA Optimus laptop.

**Comment by anup-das** [comment]
I am an ex-NVIDIA employee and I worked in their chipset division developing kernel drivers on Windows long ago. 16+ years I have developed windows kernel drivers.

I had reported a BSOD to NVIDIA  few months back caused by their GFX driver related to power events.  I could not start my laptop as it immediately crashed every time and I had to reformat my laptop and restore it with an old driver.  They could not repro but after I asked them to try out Optimus system they were able to repro. They took 1 month to fix the issue.

Based on my exp I don't think i can file this issue with them. They will redirect me back to you guys.
I wonder if there are any WHQL tests that are missing to catch such issues.
